keyboards/planck/keymaps/purple-rw/keymap.c

```diff
@@ -0,0 +1,548 @@
+),
+
+/* Plover
+ * ,-----------------------------------------------------------------------------------.
+ * |      | q S- | w T- | s P- | r H- | t *  | y *  | u -F | i -P | o -L | p -T | [ -D |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      | a S- | s K- | d W- | f R- | g *  | h *  | j -R | k -B | l -G | ; -S | ' -Z |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * | Exit |      |      | 1 #  | c A- | v O- | n -E | m -U | 1 #  |      |      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_PLOVER] = LAYOUT_planck_grid(
+    XXXXXXX, KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC,
+    XXXXXXX, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+    EXT_PLV, XXXXXXX, XXXXXXX, KC_1,    KC_C,    KC_V,    KC_N,    KC_M,    KC_1,    XXXXXXX, XXXXXXX, XXXXXXX
+),
+
+/* Adjust
+ * ,-----------------------------------------------------------------------------------.
+ * | Exit |Brght+| Sat+ | Hue+ |RGBmod| BL+  |      |      |      |      |      |Debug |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |Brght-| Sat- | Hue- | RGB  | BL-  |AGnorm|Colemk|Qwerty|HndsDn|Semimk|Plover|
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |  BL  |BLbrth|AGswap|AudOn |AudOff|MusOn |MusNxt|      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      | Exit | Exit |      |      |      |Reset |BootL |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_ADJST] = LAYOUT_planck_grid(
+    ADJST,   RGB_VAI, RGB_SAI, RGB_HUI, RGB_MOD, BL_UP,   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, DB_TOGG,
+    XXXXXXX, RGB_VAD, RGB_SAD, RGB_HUD, RGB_TOG, BL_DOWN, AG_NORM, COLEMAK, QWERTY,  HANDSDN, SEMIMAK, PLOVER,
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, BL_TOGG, BL_BRTG, AG_SWAP, AU_ON,   AU_OFF,  MU_ON,   MU_NEXT, XXXXXXX,
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, ADJST,   ADJST,   XXXXXXX, XXXXXXX, XXXXXXX, QK_RBT,  QK_BOOT
+)
+
+};
+
+#ifdef AUDIO_ENABLE
+    float plover_song[][2]     = SONG(PLOVER_SOUND);
+    float plover_gb_song[][2]  = SONG(PLOVER_GOODBYE_SOUND);
+#endif
+
+layer_state_t layer_state_set_user(layer_state_t state) {
+  return update_tri_layer_state(state, _LOWER, _RAISE, _MODS);
+}
```

Comment:
Should be using the new [TriLayer feature](https://docs.qmk.fm/#/feature_tri_layer) instead
```suggestion
```

Reply:
I must have been reading on some old materials.  Let me use this new feature.

Reply:
One issue with the new Tri Layer feature is that there can be only one "tri layer".  Not a problem here, but it could be a problem in the future when I want to add some macro layers.

The need of defining a number in config.h seems to be an issue, as layers are enum's in keymap.c.  If I add/remove layers in keymap.c, I need to update the numbers in config.h.  Not ideal.

A real usability issue is that I can't lock the layer like TT(_RAISE), with several quick taps.  Any suggestion?

Reply:
>  Not a problem here

You said it already, not an issue, let's worry about it when (if) it appears

> The need of defining a number in config.h seems to be an issue

An `enum` is nothing but a fancy way of defining numbers, with your current code `_COLEMAK` is effectively a `0` to the code, `_QWERTY` is `1` and so on... 
Since the feature also provides setter methods, you dont **need** to configure this at compile time, you can perfectly use these names to configure the feature as soon as your board starts, i.e. add to your keymap:
```c
void keyboard_post_init_user(void) {
    set_tri_layer_lower_layer(_NAVFUNC);
    set_tri_layer_upper_layer(_SYMBNUM);
    set_tri_layer_adjust_layer(_MODS);
}
```
> A real usability issue is that I can't lock the layer like TT(_RAISE), with several quick taps. Any suggestion?

How so? What leads you to think `TT(_RAISE)` wont work? The fact that TriLayer feature provides you a keycode to turn that layer on/off (effectively a `MO`) and that a 3rd layer is controlled by the state of the "lower" and "upper" layers doesn't restrict you from controlling those two layers in another way

Reply:
for the defines, you'd need to use the number, not the name.  both ways amount to the same thing, but the keyboard_post_init_user method is a bit more human readable.

Reply:
> How so? What leads you to think TT(_RAISE) wont work? The fact that TriLayer feature provides you a keycode to turn that layer on/off (effectively a MO) and that a 3rd layer is controlled by the state of the "lower" and "upper" layers doesn't restrict you from controlling those two layers in another way

I mean, without the new tri layer function, I would put `TT(_RAISE)` for the RAISE layer key.  But with tri layer, I need `TL_LOWR`.  How do I make it so that TL_LOWR would lock the lower layer when I quickly tap it couple (3) times?  The only way I can think of right now is to implement it in process_record_user().  A simpler way is preferred.